dbms/src/DataStreams/DistinctBlockInputStream.cpp

```diff
@@ -122,11 +92,59 @@ Block DistinctBlockInputStream::readImpl()
 
 bool DistinctBlockInputStream::checkLimits() const
 {
-	if (max_rows && set.size() > max_rows)
+	if (max_rows && data.getTotalRowCount() > max_rows)
 		return false;
-	if (max_bytes && set.getBufferSizeInBytes() > max_bytes)
+	if (max_bytes && data.getTotalByteCount() > max_bytes)
 		return false;
 	return true;
 }
 
+template <typename Method>
+void DistinctBlockInputStream::buildFilter(
+	Method & method,
+	const ConstColumnPlainPtrs & columns,
+	IColumn::Filter & filter,
+	size_t rows) const
+{
+	typename Method::State state;
+	state.init(columns);
+
+	/// Для всех строчек
+	for (size_t i = 0; i < rows; ++i)
+	{
+		/// Строим ключ
+		typename Method::Key key = state.getKey(columns, columns.size(), i, key_sizes);
+
+		/// Если вставилось в множество - строчку оставляем, иначе - удаляем.
+		filter[i] = method.data.insert(key).second;
+
+		if (limit && data.getTotalRowCount() == limit)
```

Comment:
Let put it outside of the loop (so, remove).
Limits are checked only at each block, not in inner loops, everywhere else.